ludwig/benchmarking/reporting.py

```diff
@@ -1,35 +1,166 @@
-import os
-from typing import Any, Dict, Tuple
-
-from ludwig.constants import CACHE, EVAL_TAG, EXPERIMENT_RUN, TRAIN_TAG
-from ludwig.utils.data_utils import load_json, save_json
-from ludwig.utils.misc_utils import merge_dict
-
-
-def create_metrics_report(experiment_name: str) -> Tuple[Dict[str, Any], str]:
-    """Compiles performance and non-performance metrics.
-
-    `experiment_name`: name referring to the experiment.
-    Returns a full report and the path where it's saved.
-    """
-    full_report = dict()
-    os.makedirs(os.path.join(os.getcwd(), experiment_name, "metrics_report"), exist_ok=True)
-    for tag in [TRAIN_TAG, EVAL_TAG]:
-        if tag == TRAIN_TAG:
-            resource_usage_path = os.path.join(os.getcwd(), experiment_name, CACHE, "train_resource_usage_metrics.json")
-            performance_path = os.path.join(os.getcwd(), experiment_name, EXPERIMENT_RUN, "training_statistics.json")
-        elif tag == EVAL_TAG:
-            resource_usage_path = os.path.join(
-                os.getcwd(), experiment_name, CACHE, "evaluate_resource_usage_metrics.json"
-            )
-            performance_path = os.path.join(os.getcwd(), experiment_name, EXPERIMENT_RUN, "test_statistics.json")
+from collections import Counter, defaultdict
+from typing import Any, Dict, List, Tuple, Union
+
+import torch
+from torch._C._autograd import _KinetoEvent
+from torch.autograd import DeviceType, profiler_util
+
+from ludwig.constants import LUDWIG_TAG
+
+
+def initialize_stats_dict(main_function_events: List[profiler_util.FunctionEvent]) -> Dict[str, Any]:
+    """Initialize dictionary which stores resource usage information per tagged code block.
+
+    :param main_function_events: list of main function events.
+    """
+    info = dict()
+    for event_name in [evt.name for evt in main_function_events]:
+        info[event_name] = []
+    return info
+
+
+def get_memory_details(kineto_event: _KinetoEvent) -> Tuple[str, int]:
+    """Get device name and number of bytes (de)allocated during an event.
+
+    :param kineto_event: a Kineto event instance.
+    """
+    if kineto_event.device_type() in [DeviceType.CPU, DeviceType.MKLDNN, DeviceType.IDEEP]:
+        return "cpu", kineto_event.nbytes()
+    elif kineto_event.device_type() in [DeviceType.CUDA, DeviceType.HIP]:
+        return f"cuda_{kineto_event.device_index()}", kineto_event.nbytes()
+    else:
+        raise ValueError(f"Device {kineto_event.device_type()} is not valid.")
+
+
+def get_device_memory_usage(
+    kineto_event: _KinetoEvent, mem_records_acc: profiler_util.MemRecordsAcc, run_usage_info: Dict[str, Any]
+) -> Dict[str, Any]:
```

Comment:
Dataclass?

Reply:
The size and keys of this dictionary will depend on the devices that the code is run on. The contents will mainly differ based on the number of GPUs.